proposals/0000-binaryinteger-ispower.md

```diff
@@ -0,0 +1,271 @@
+
+Apart from the *is-power-of-two* usage, queries on *non-2* bases may also be practical, though they are much less common. In signal processing, for example, the efficient radix-4 algorithms can be applied when the FFT size is a power of 4.
+
+As a result, it would be beneficial if we could have such an API supported in the standard library.  To be more specific, it is an extension method in the form of `isPower(of:)` to the `BinaryInteger` protocol, checking whether the `BinaryInteger` *self* is a power of the base specified by the parameter.  Let us discuss the impacts in the following aspects:
+
+### Readability
+
+A classic approach to check if an integer `n` is power of two is to test whether the condition `n > 0 && n & (n - 1) == 0` holds. Although such code is efficient, its underlying functionality is not intuitive to many people. As a result, programmers would have to put additional information somewhere (e.g. in the comments nearby) to make it clear.
+
+Below is an example when making an assertion that `bucketCount` is power of two.  The classic approach is applied, followed by an additional error message for necessary clarification purpose.  However, if we had the proposed API available, the code would become more fluent and concise.
+```swift
+// Example (1) - apple/swift/stdlib/public/core/HashTable.swift
+internal struct _HashTable {
+    internal init(words: UnsafeMutablePointer<Word>, bucketCount: Int) {
+        _internalInvariant(bucketCount > 0 && bucketCount & (bucketCount - 1) == 0,
+          "bucketCount must be a power of two")
+
+//      _internalInvariant(bucketCount.isPower(of: 2)) --- proposed solution
+
+        ...
+    }
+}
+```
+
+### Efficiency
+
+The user-implemented code may be less performant (e.g. [#1](https://github.com/OpsLabJPL/MarsImagesIOS/blob/f2109f38b31bf1ad2e7b5aae6916da07d2d7d08e/MarsImagesIOS/Math.swift#L17), [#2](https://github.com/eaheckert/Tournament/blob/c09c6b3634da9b2666b8b1f8990ff62bdc4fd625/Tournament/Tournament/ViewControllers/TCreateTournamentVC.swift#L215)),  since some developers are not aware of the classic approach as described above.
+
+The example below shows a controversial approach, which employs the `llvm.ctpop` intrinsic to count the number of 1 bits. It can be expensive when the hardware does not have relevant `popcount` instruction support.
```

Comment:
1 must be singular

```suggestion
The example below shows a controversial approach, which employs the `llvm.ctpop` intrinsic to count the number of 1 bit. It can be expensive when the hardware does not have relevant `popcount` instruction support.
```